### DISCUSSION TITLE
Takes a really long time before headlessJS task is fired in background

### DISCUSSION THREAD

**Discussion by CaptainJeff** [invertase/react-native-firebase#4470] [Q&A] [answered]
I am receiving my headless js background task through `messaging().setBackgroundMessageHandler` and looking at logcat it appears as though the message is received by firebase almost immediately but then the app instantiates (which takes 20/25 seconds) before my headless js setBackgroundMessageHandler function is called. Any thoughts here? How can I speed this up or call my headless js task before the rest?

**Comment by mikehardy** [comment, marked as answer]
Hi there! This *seems* like it is going to be project-specific performance issue, I would throw the standard suite of performance / tracing tools at it from AndroidStudio / firebase performance (to trace startup paths' time taken) and/or console output at various startup points to see exactly what is taking so long. Might also be worth (as a test) stripping everything out except index.js / App.js in order to get a "nothing included" / "fastest possible" baseline. I don't recall seeing this in my app, for what it's worth

**Reply by CaptainJeff** [reply]
@mikehardy thanks for your response, as always! I'll start striping away and see if I can find the cause. I have a suspicion its happening when i instantiate react-navigation / mobx

I'll report back when I have anything

**Reply by CaptainJeff** [reply]
@mikehardy it seems like the bottleneck is mostly coming from dependencies and require cycles that I needed to improve. Thanks for your response

**Reply by TommyLeong** [reply]
Hello @CaptainJeff, what's your current response time from `messaging().setBackgroundMessageHandler` after resolving your bottleneck?

**Reply by CaptainJeff** [reply]
@TommyLeong I still havent' fully resolved this yet actually because I need some of logic that I had commented out to test the issue. 

When I originally made this post it was roughly 20s on a debug build. With everything commented out and fixing require cycles I probably cut that down to 11s. And now i'm at about 12s or 13s by including some of that logic back in.

**Reply by TommyLeong** [reply]
@CaptainJeff  Thanks for sharing 🙏
